### PR TITLE
Fix for CAMEL-8287

### PR DESCRIPTION
**Pull request by hutchkintoot**
Allows a user to supply a custom URIResolver to be used to resolve imports in his/her schematron rules files.

### CODE REVIEW

**Comment by akhettar** [comment]
@davsclaus can you merge this pull request as I don't have admin right on master branch? I have reviewed it, all look good to me thanks.

File: components/camel-schematron/src/main/java/org/apache/camel/component/schematron/processor/ClassPathURIResolver.java [hunk 1]
```diff
@@ -28,17 +29,33 @@
...
     }
 
     @Override
     public Source resolve(String href, String base) throws TransformerException {
-        return new StreamSource(ClassPathURIResolver.class.getClassLoader()
-                .getResourceAsStream(rulesDir.concat(File.separator).concat(href)));
+        InputStream stream = ClassPathURIResolver.class.getClassLoader()
+                .getResourceAsStream(rulesDir.concat(File.separator).concat(href));
+
+        if (null != stream) {
```
Comment (davsclaus) [inline components/camel-schematron/src/main/java/org/apache/camel/component/schematron/processor/ClassPathURIResolver.java]: Flip the if as we do X != null style in Camel

File: components/camel-schematron/src/test/java/org/apache/camel/component/schematron/processor/SchematronProcessorTest.java [hunk 2]
```diff
@@ -16,8 +16,8 @@
  */
 package org.apache.camel.component.schematron.processor;
 
-import javax.xml.transform.Templates;
-import javax.xml.transform.TransformerFactory;
+import javax.xml.transform.*;
```
Comment (davsclaus) [inline components/camel-schematron/src/test/java/org/apache/camel/component/schematron/processor/SchematronProcessorTest.java]: W do not use * imports. Run: mvn compile -Psourcecheck to run checkstyle to see what problems you have

Comment (hutchkintoot) [inline components/camel-schematron/src/test/java/org/apache/camel/component/schematron/processor/SchematronProcessorTest.java]: Sorry, will fix!

**Comment by hutchkintoot** [comment]
mvn compile -Psourcecheck now gives green light and null-checks flipped.

File: components/camel-schematron/src/main/java/org/apache/camel/component/schematron/SchematronEndpoint.java [hunk 3]
```diff
@@ -112,6 +116,11 @@ public void setRules(Templates rules) {
         this.rules = rules;
     }
 
+    public void setUriResolver(URIResolver uriResolver) { this.uriResolver = uriResolver; }
```
Comment (davsclaus) [inline components/camel-schematron/src/main/java/org/apache/camel/component/schematron/SchematronEndpoint.java]: You need to add javadoc to document this option on the setter.

File: components/camel-schematron/src/test/resources/custom-resolver/title-rules.sch [hunk 4]
```diff
@@ -0,0 +1,3 @@
+<sch:rule context="title" xmlns:sch="http://purl.oclc.org/dsdl/schematron">
```
Comment (davsclaus) [inline components/camel-schematron/src/test/resources/custom-resolver/title-rules.sch]: Can we add a license header to this file?

**Comment by hutchkintoot** [comment]
Javadoc and license header in place.

**Comment by davsclaus** [comment]
The PR *must* be on master only. Camel 2.15.x is EOL.

Please close this PR and create a new PR against master branch.

**Comment by hutchkintoot** [comment]
New pull request sent. Now based on master. Would really need this back ported to 2.16.x and/or 2.17.x if possible. We are currently running on 2.15.3.

**Pull request closed without merging**